### ISSUE TITLE
Revise pmix_key_t to reduce memory footprint

### ISSUE THREAD

**Issue by rhc54** [pmix/pmix-standard#367]
### Overview

PMIx information is frequently stored as `pmix_info_t` structures. The size of these structures is driven by the `pmix_key_t` definition as a fixed-size character array. This RFC focuses on discussion whether or not to reduce the size of this definition, and if so, then by what means.

### Motivation

The Standard currently defines the size of `pmix_key_t` as 512 bytes (PMIX_MAX_KEYLEN of 511 plus one location for the NULL terminator). Currently, the Standard only contains keys that are _much_ smaller than this, typically in the 10-character range. While it is true that user-defined keys can frequently be larger, it is unusual for such keys to exceed several tens-of-bytes in length in practice.

Reducing the memory footprint of implementations would be aided by changing the definition of `pmix_key_t` to a `char*` pointer, thus minimizing the storage required by the` pmix_info_t`. The rationale behind the initial definition was to ensure that a shared memory representation of an array of `pmix_info_t` would have a fixed step size between elements. However, this could be maintained by ensuring that the string in the `pmix_key_t` not be inlined - i.e., that the `char*` pointer not point to a string contained within the `pmix_info_t` itself. This is an implementation-level decision, but is noted here to explain why the described change could be made without violating the initial rationale.

The question for discussion therefore is: does the Standard want to revise the definition of `pmix_key_t`? If so, then what should be done - reduce the size of the fixed array, change it to a `char*` pointer, or...?

**Comment by raffenet** [comment]
Note from Nov 2021 meeting: In OpenPMIx today, using `char *` is probably the best bet. It may cause a performance hit, but might not be significant enough to worry about. Plan is to write it up as a proposal and gather feedback. This change would likely be an ABI break for existing implementations.

**Comment by jjhursey** [comment]
I think we should entertain moving away from the fixed-length array for the key length and instead move to something that lends itself to more flexible representations. This will be an ABI break, but something to do now that we are first defining the ABI.

I have a couple of options to consider.

## Current

Currently, we have (taking these from OpenPMIx `pmix_common.h` for simplicity):
```
#define PMIX_MAX_KEYLEN    511

typedef char pmix_key_t[PMIX_MAX_KEYLEN+1];
typedef uint32_t pmix_info_directives_t;
typedef struct pmix_value {
    pmix_data_type_t type;
    union {
        ...
        pmix_nspace_t *nspace; // Notice this is a pointer, not a fixed array
    } data;
} pmix_value_t;

typedef struct pmix_info {
    pmix_key_t key;
    pmix_info_directives_t flags;
    pmix_value_t value;
} pmix_info_t;

typedef struct pmix_pdata {
    pmix_proc_t proc;
    pmix_key_t key;
    pmix_value_t value;
} pmix_pdata_t;

typedef struct pmix_regattr_t {
    char *name;
    pmix_key_t string;
    pmix_data_type_t type;
    char **description;
} pmix_regattr_t;
```

Assorted notes:
 * The focus of the discussion is on the `pmix_value_t`, but there are a few other structures that are also impacted (See above). The `pmix_value_t` is probably the most widely used though.
 * (-) For every value structure being used by the client we have `512` characters, often not used.
 * (-) Comparing keys requires a string operation.
 * (+) String representation allows the client flexibility in describing the value being passed around. Note that key/value pairs are often used for bookmark exchange where the client decides on the key names. So they are not just those defined as reserved keys by PMIx.
 * (+) We already have `PMIX_VALUE_CONSTRUCT`, `PMIX_VALUE_DESTRUCT`, `PMIX_VALUE_FREE`. That we could leverage.
 * (+) Not changing it doesn't break any existing PMIx consumers

## Option 1: Replace fixed array with string pointer

```
typedef char* pmix_key_t;

typedef struct pmix_info {
    pmix_key_t key;
    pmix_info_directives_t flags;
    pmix_value_t value;
} pmix_info_t;
```

Assorted notes:
 * (+) Only the space required for the key string is allocated by the PMIx implementation
 * (+) Keys can be any length, but must be NULL terminated.
 * (-) Who frees the memory?

On free'ing memory, if the user allocates the key then they are responsible for freeing it. If the PMIx library allocated it then the user will have to call a `PMIX_KEY_DESTROY` function or free the memory before it goes out of scope.

I think if we define a rule for who is responsible for the memory allocated then we will be ok here. That gives maximum flexibility to consumers of PMix on how they name their keys to both avoid conflicts and be meaningful to their libraries.

## Option 2: Replace fixed array with an integer

```
typedef uint64_t pmix_key_t;

typedef struct pmix_info {
    pmix_key_t key;
    pmix_info_directives_t flags;
    pmix_value_t value;
} pmix_info_t;
```

Assorted notes:
 * (+) We could reserve a 'high' bit to designate PMIx reserved vs other keys
 * (+) Don't have to worry about free'ing memory
 * (+) No string comparison operations on the key
 * (-) Application will need to define a isolated set of integers that will not conflict with other consumers of the PMIx interface.

I propose this because of the symmetry with my proposal on Issue #366 . However, I don't know if it works well for keys since the user can also define keys (unlike namespaces which are almost exclusively defined by PMIx).

It will be common that the application contains multiple libraries each of which may be exchanging key/value information via PMIx. Currently, each library could prefix their keys to make them distinct from one another (e.g., `mylib.XYZ`). If we move to a more restricted integer range then the libraries won't have as easy of a mechanism to avoid one another. They would have to agree on available ranges. We could hash the string representation of they keys into the 64 bit space, but now we would have to manage conflicts which may get tricky and messy.

For that reason I don't think replacing the key with an integer would be a great option for the key.

**Comment by jjhursey** [comment]
## Option 3: Replace fixed array with an integer via string registration

```
typedef uint64_t pmix_key_t;

// Local operations. ID's meaningful only to the local client
int pmix_key_register(char * key_name, pmix_key_t *key_id);
int pmix_key_deregister(pmix_key_t key_id)
int pmix_key_get_string(pmix_key_t key_id, char ** key_name);

typedef struct pmix_info {
    pmix_key_t key_id;
    pmix_info_directives_t flags;
    pmix_value_t value;
} pmix_info_t;
```

Assorted notes:
 * (+) We could reserve a 'high' bit to designate PMIx reserved vs other keys
 * (+) Don't have to worry as much about free'ing memory. Memory ownership is tied to the semantics of `pmix_key_*` functions
 * (+) No string comparison operations on the key 'id'
 * (+) Local registration of a string that PMIx can check for conflicts. Conflict check would be a new enhancement. Currently, if library A uses the same key as library B then they will walk over each others keys.
 * (-) Additional API complexity

This would split the difference between Option 1 and 2. It allows the user to create arbitrary string representations of keys, and still only pass around an integer reference to that key.

Semantically, the integer reference for the key is valid only for the local client. The PMIx implementation can decide if it wants to make the integer reference globally unique (incurring overhead for doing so). If it is only locally unique then the PMIx implementation will need to pass around the string representation (since ids might be assigned differently at different nodes). If the PMIx implementation chooses to make them globally unique then they could just pass the integer id around instead of the string representation. This is a PMIx implementation optimization. The client should not assume that the id is globally unique across PMIx implementation instances.

**Comment by rhc54** [comment]
I've spent some time now digging into the impact of making such a change - and have concluded that it isn't worth pursuing (regardless of option). The impact on existing users would potentially be large and there is no real benefit to the implementation in terms of memory footprint or lookup performance.

If users utilize the PMIx macros for all key-related operations, then we can protect them. We simply alter the macro to accommodate the changes to `pmix_info_t` and the other structures. However, there is no rule requiring all access to be via the macros, and some people have inevitably chosen to directly access those values. In these cases, user code would have to be revised and protected against the difference in PMIx versions.

When looking at the internals of the existing implementation, I found that we already transfer the user-provided keys to non-fixed arrays. It is easy enough for us to internally create a table and save most of that memory as well by replacing the resulting string with an integer table index in our implementation (which also gives us the lookup performance boost). We don't require any change to the public-facing Standard for this purpose.

I would therefore recommend that we close this issue and leave the `pmix_key_t` definition as-is.

**Comment by rhc54** [comment]
I've been coding a prototype of this change in order to quantify the impact. After reworking OpenPMIx, I find that:

* there is no impact I could detect on API signatures
* anyone who uses the macros is largely protected from the change. Modifications to the macros aren't terribly intrusive and fully "hide" the change in key type
* anyone who did not use the macros is likely to encounter trouble. The current definition guarantees a static allocation of memory for the key's string. Changing the key to a `char*` removes that guarantee, thereby leading to potential segfaults - e.g., doing a `strncpy` into a `pmix_key_t` will now fail.

Compilers sometimes help to catch these problems by generating "not initialized" warnings, but this does not provide complete protection. In fact, most of the changes I had to make to OpenPMIx stemmed from directly using `strncpy` instead of the PMIx macros.

Bottom line: we can make this work, BUT it will likely force source-code modifications by current users, particularly by those who did not use the macros (e.g., people who `dlopen` the library instead of linking against it). This may be alleviated once the `pmix-abi` headers are released, and is a one-time change. However, there could be issues when building across versions - e.g., if the app is built against a "new" definition for `pmix_key_t` and then executed (via `dlopen`) against a library with an "old" definition (or vice versa). Hence, the code would probably require version protection, which could get messy.

**Comment by jjhursey** [comment]
Note from IA WG meeting on Jan. 24, 2022 (revisited on March 14, 2022):
 * Inside the PMIx library the string associated with the `pmix_key_t` can be handled in a more flexible manner than how it is exposed to the user.
 * I think Ralph did a good job of summarizing the core discussion in the [comment above](https://github.com/pmix/pmix-standard/issues/367#issuecomment-1017079248)
 * **Decision:** It was the decision of the working group to leave the `pmix_key_t` as it is today. If additional uses cases are brought forward then we will certainly reconsider.

Per the notes above I'm going to close this ticket. We can reopen it if further discussion is desired.
